### DISCUSSION TITLE
Customised fedora cloud image not completing the startup on Linux (fedora host).

### DISCUSSION THREAD

**Discussion by Ranjandas** [lima-vm/lima#2335] [Q&A] [answered]
I am using Packer to customise the Fedora 40 cloud image. The customised image is launched using a custom lima template.

This works fine on macOS (with `vmType:` both  `vz` and `qemu`). However, when I did the same on a fedora host, lima was stuck at the following stage:

```
...
INFO[0003] SSH Local Port: 42575
INFO[0002] [hostagent] Waiting for the essential requirement 1 of 2: "ssh"
INFO[0012] [hostagent] Waiting for the crucial requirement 1 of 2: "ssh"
...
...
```
When I checked the `serial.log`,  I could see that the cloud-init failed to create my user account.

```
[    7.825978] cloud-init[910]: 2024-05-10 13:57:32,482 - schema.py[WARNING]: Invalid cloud-config provided: Please run 'sudo cloud-init schema --system' to see the schema errors.
[    9.564942] BTRFS info (device vda4): resize device /dev/vda4 (devid 1) from 4212109312 to 106217582592
[   10.602674] cloud-init[910]: 2024-05-10 13:57:35,257 - util.py[WARNING]: Failed to create user ranjan
[   10.611336] cloud-init[910]: 2024-05-10 13:57:35,267 - util.py[WARNING]: Running module users_groups (<module 'cloudinit.config.cc_users_groups' from '/usr/lib/python3.12/site-packages/cloudinit/config/cc_users_groups.py'>) failed
[   10.840111] cloud-init[910]: Generating public/private rsa key pair.
```

Any hint on where to look for the issue? I'm just wondering why it fails only on Linux.

**Comment by afbjorklund** [comment]
Are there any details in `/var/log/cloud-init.log`?

**Reply by Ranjandas** [reply]
I am unable to access that file because I can't get inside the VM, as the user provisioning itself fails. Interestingly, the user account I created as part of the packer build is also not working.

Btw, this is the `user-data` that I am using (passing as `-cdrom cidata.iso`) with the packer build to create a user account . Would this in any way conflict with the user-data that lima is using (specifically in lLima on Linux hosts, as this whole stuff works on macOS). 

```
#cloud-config

system_info:
  default_user:
    name: fedora

chpasswd:
  list: |
    fedora:fedora
  expire: False
ssh_pwauth: True
resize_rootfs: True
```

**Reply by afbjorklund** [reply]
Possibly, I don't think the lima start is able to handle multiple cloud-init images (cidata.iso) even if default users are common

But it is just guesses, until we see the error. You can run with a display to access the console, or use the `serial.sock` created

**Reply by Ranjandas** [reply, marked as answer]
Thank you for your inputs @afbjorklund. I managed to get into the VM using `serial.sock` and could inspect the cloud-init.log.

Looks like it is failing because the `useradd` command is using UID 1000 (from my linux host), which is already taken by my `fedora` user.

```
2024-05-13 06:49:13,815 - __init__.py[DEBUG]: Adding user ranjan
2024-05-13 06:49:13,815 - subp.py[DEBUG]: Running hidden command to protect sensitive input/output logstring: ['useradd', 'ranjan', '--home', '/home/ranjan.linux', '--shell', '/bin/bash', '--uid', '1000', '-m']
2024-05-13 06:49:13,943 - util.py[WARNING]: Failed to create user ranjan
2024-05-13 06:49:13,943 - util.py[DEBUG]: Failed to create user ranjan
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/cloudinit/distros/__init__.py", line 742, in add_user 
    subp.subp(useradd_cmd, logstring=log_useradd_cmd)
  File "/usr/lib/python3.12/site-packages/cloudinit/subp.py", line 322, in subp
    raise ProcessExecutionError(
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['useradd', 'ranjan', '--home', '/home/ranjan.linux', '--shell', '/bin/bash', '--uid', '1000', '-m']
Exit code: 4
Reason: -
Stdout: 
Stderr: useradd: UID 1000 is not unique
2024-05-13 06:49:13,953 - handlers.py[DEBUG]: finish: init-network/config-users_groups: FAIL: running config-users_groups with frequency once-per-instance
2024-05-13 06:49:13,954 - util.py[WARNING]: Running module users_groups (<module 'cloudinit.config.cc_users_groups' from '/usr/lib/python3.12/site-packages/cloudinit/config/cc_users_groups.py'>) failed
2024-05-13 06:49:13,954 - util.py[DEBUG]: Running module users_groups (<module 'cloudinit.config.cc_users_groups' from '/usr/lib/python3.12/site-packages/cloudinit/config/cc_users_groups.py'>) failed 
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/cloudinit/config/modules.py", line 255, in _run_modules
    ran, _r = cc.run(
              ^^^^^^^
  File "/usr/lib/python3.12/site-packages/cloudinit/cloud.py", line 60, in run
    return self._runners.run(name, functor, args, freq, clear_on_fail)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/cloudinit/helpers.py", line 172, in run
    results = functor(**args)
              ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/cloudinit/config/cc_users_groups.py", line 240, in handle
    cloud.distro.create_user(user, **config)
  File "/usr/lib/python3.12/site-packages/cloudinit/distros/__init__.py", line 800, in create_user
    self.add_user(name, **kwargs)
  File "/usr/lib/python3.12/site-packages/cloudinit/distros/__init__.py", line 745, in add_user
    raise e
  File "/usr/lib/python3.12/site-packages/cloudinit/distros/__init__.py", line 742, in add_user
    subp.subp(useradd_cmd, logstring=log_useradd_cmd)
  File "/usr/lib/python3.12/site-packages/cloudinit/subp.py", line 322, in subp
    raise ProcessExecutionError(
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['useradd', 'ranjan', '--home', '/home/ranjan.linux', '--shell', '/bin/bash', '--uid', '1000', '-m']
Exit code: 4
Reason: -
Stdout: 
Stderr: useradd: UID 1000 is not unique
```

Got it working by updating my packer cloud-init to have a different UID.

```
#cloud-config

system_info:
  default_user:
    name: fedora
    uid: 1001 <------ change the UID to avoid conflict with the UID of the host user account

chpasswd:
  list: |
    fedora:fedora
  expire: False
ssh_pwauth: True
resize_rootfs: True
```

Figured out that lima gets the user information from the host here: https://github.com/lima-vm/lima/blob/7e493aa20d24161b3ed65a14bb72cec8516824e5/pkg/osutil/user.go#L112

Everything is working fine now. Thank you once again :pray:

**Reply by afbjorklund** [reply]
Default passwords are considered bad security practice, but Ubuntu Server is also doing this (unlike Cloud without it)

Lima is using keys for access, so disables password by default. You do have to add passwords, to access through GUI

**Reply by Ranjandas** [reply]
Thank you, @afbjorklund; yes, I am aware of the security implications. I have replaced the password with `ssh_authorized_keys`. 

The new `user-data` is:

```
#cloud-config

system_info:
  default_user:
    name: fedora
    uid: 1001

resize_rootfs: True
ssh_authorized_keys:
  - ssh-rsa AAAAB3Nza....
```

**Reply by afbjorklund** [reply]
I don't understand why you need a default user in the image, instead of adding one during instantiation

i.e. why not just use `template://fedora`

**Reply by Ranjandas** [reply]
@afbjorklund, I was doing that initially, but the package installations during the provisioning stage added to the startup time (especially when spawning a multi-node setup), so I decided to bake custom images and launch them using Lima. Currently, I am using this `--set` option to launch the VM from the output of packer builds.

```
## the vm.yaml doesn't have images[] in it by default
$ limactl start vm.yaml --set='. |= .images[0].location="file://...
```

**Reply by afbjorklund** [reply]
Adding packages is understandable,  my question was for the user account.

**Reply by Ranjandas** [reply]
Sorry! I don't think I understood the question on the user account. There should be a user account for my packer provisioner to run, right? Are you asking specifically about the use of `default_user` instead of maybe `user` module? 🤔

**Reply by Ranjandas** [reply]
Ah! I think I now understand your question. The reason for me to have a user account is:

* I am using Packer SSH provisioner
* I have a [`shutdown_command`](https://developer.hashicorp.com/packer/integrations/hashicorp/qemu/latest/components/builder/qemu#shutdown-configuration) specified to gracefully shutdown the VM
* I can also customize the script using custom environment variables, etc, which will be difficult if I go via the user-data path.

If I had fully depended on the cloud-init for package installation etc, I could have avoided creating the `fedora` account. I might give it a shot and see if it fulfils my needs.

Is my understanding correct? I think there is a challenge with that: Packer won't be able to detect when the cloud-init finishes before shutting down the temporary VM.

**Reply by afbjorklund** [reply]
Ok, so it is more related to the Packer tool (no longer Open Source) than about the resulting 

My question was why you needed to add a user account for Lima, since it already does that.
